options/lsb/generic/tls.cpp

```diff
@@ -2,11 +2,17 @@
 #include <mlibc/thread.hpp>
 #include <mlibc/rtdl-abi.hpp>
 
-struct __abi_tls_entry;
-
-extern "C" void *__dlapi_get_tls(struct __abi_tls_entry *);
-
-extern "C" void *__tls_get_addr(struct __abi_tls_entry *entry) {
-	return __dlapi_get_tls(entry);
-}
+#if defined(__riscv) && defined(MLIBC_STATIC_BUILD)
+	// On RISC-V, linker optimisation is not guaranteed and so we may still get
+	// calls to this function in statically linked binaries.
+	extern "C" void *__tls_get_addr(struct __abi_tls_entry *entry) {
+		Tcb *tcbPtr = mlibc::get_current_tcb();
+		auto dtvPtr = reinterpret_cast<char *>(tcbPtr->dtvPointers[0]);
+		return reinterpret_cast<void *>(dtvPtr + entry->offset);
+	}
+#else
+	extern "C" void *__tls_get_addr(struct __abi_tls_entry *entry) {
+		return __dlapi_get_tls(entry);
+	}
+#endif
```

Comment:
IIRC it uses `entry->object` which hasn't been set up properly (since we don't perform dynamic relocs), in the static case.

Reply:
Do relocations exist that we could perform to fix this? This `#ifdef` approach breaks if a statically linked binary `dlopen()`s a shared library at runtime.

Reply:
Does dlopen usually work from static binaries? The DSO would have to be linked against libc statically too (or not at all)

Reply:
I guess it can be made to work but OTOH it might be more reasonable to merge this PR now; we can always worry about this later. Maybe add a comment that the current implementation fails if dlopen() is used from a static binary.